### PR TITLE
Further Support Null Analysis

### LINKED ISSUE #43: JSpecify 0.3.0, Checker Framework nullness, and avaje-record-builder in Maven
Or, propagating TYPE_USE annotations (#19) is insufficient.

Or, builders are really validators.

---

I've been fooling around with nullability analysis to get a sense of the state of the ecosystem primo 2024. For this experiment I'm using

* **Maven**, because JetBrains claims 60% of Java projects use Maven and because I hate it less than Gradle.
* **Checker Framework's nullness checker**, because it's well established, can run without a third-party host, and NullAway depends on it anyway.
* **JSpecify**, because it's ~the only _general_ TYPE_USE nullability annotation library and it specifically lacks certain tool specific support.
* **avaje-record-builder**, because avaje-record-builder has some recognition of TYPE_USE annotations, and anyway, because I suspect any dialogue about the state of third-party nullability analysis in combination with a record builder generator has a better chance of being fruitful here no matter the outcome.

I'm filing this issue for posterity to record that this combination of tools presently is not workable. The essence of why is the same as the essence of https://github.com/uber/NullAway/issues/121, the outcome of which is not clear to me. I am aware of the extensive history represented by #19 and Randgalt/record-builder#111.

There are variations I have not experimented with much or at all and so cannot really comment on:

* Checker Framework standalone execution, which offers flexible control of what to analyse when and possibly trivially allows for ignoring problematic elements like generated code.
* Checker Framework via Gradle, but I get the sense that Checker Framework pushes Gradle hard and that the Gradle integration is richer than the Maven integration.
* NullAway, which itself is much laxer than Checker Framework, and via Error Prone offers flexible analysis inclusion and exclusion rules even in Maven integration.
 * io.soabase.record-builder:record-builder:38, which is unable to recognize TYPE_USE annotations.

Given a record

```java
package example; // avaje/avaje-record-builder#41

@io.avaje.recordbuilder.RecordBuilder
public record Nullity(
        @org.jspecify.annotations.NonNull String a,
        @org.jspecify.annotations.Nullable String b) {}
```

avaje-record-builder-1.0 generates the moral equivalent of

```java
package example;

import io.avaje.recordbuilder.Generated;
import java.util.function.Consumer;
import org.jspecify.annotations.NonNull;
import org.jspecify.annotations.Nullable;

@Generated("avaje-record-builder")
public class NullityBuilder {
  private String a;
  private String b;

  private NullityBuilder() {} // initialization.fields.uninitialized

  private NullityBuilder(@NonNull String a, @Nullable String b) {
    this.a = a;
    this.b = b; // assignment
  }

  public static NullityBuilder builder() { return new NullityBuilder(); }

  public static NullityBuilder builder(Nullity from) { return new NullityBuilder(from.a(), from.b()); }

  public Nullity build() { return new Nullity(a, b); }

  public NullityBuilder a(@NonNull String a) {
      this.a = a;
      return this;
  }

  public NullityBuilder b(@Nullable String b) {
      this.b = b; // assignment
      return this;
  }
}
```

When compiled (POM below), Checker Framework produces three errors:

```
[ERROR] /.../src/avaje-record-builder-nullity/target/generated-sources/annotations/example/NullityBuilder.java:[14,10] error: [initialization.fields.uninitialized] the constructor does not initialize fields: a, b
[ERROR] /.../src/avaje-record-builder-nullity/target/generated-sources/annotations/example/NullityBuilder.java:[19,13] error: [assignment] incompatible types in assignment.
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
[ERROR] /.../src/avaje-record-builder-nullity/target/generated-sources/annotations/example/NullityBuilder.java:[51,15] error: [assignment] incompatible types in assignment.
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
```

The line numbers refer to the canonical generated lines, not the compressed version shown above. The compressed version is annotated with the error codes.

There are two errors in the generated code, one "obvious" and one subtle. Other limitations in Checker Framework, Checker Framework's Maven integration, and avaje-record-builder combine to make those two errors somewhere between difficult and impossible to work around (I have not found a way).

1. `assignment`: Although the `NullityBuilder b(@Nullable String b)` receiver correctly had `org.jspecify.annotations.Nullable` propagated from the `Nullity.b` component the annotation was not propagated to the `NullityBuilder.b` field, which Checker Framework consequently resolves to `@NonNull`. `@Nullable` should have been propagated all the way to the field.
   * This is the "obvious" error as far as Checker Framework's rules go.
3. `initialization.fields.uninitialized`: The `builder()` factory instantiates a builder in its default state according to the JLS, i.e. with all fields uninitialized and implicitly null. This is valid Java but Checker Framework effectively resolves those fields to `@NonNull`, which is a contradiction and therefore rejected. This behaviour is consistent with the various rules in play but the rules that are in play are not representative of a builder's role. The trouble is that `NullityBuilder.[@o.j.a.NonNull] String a` is an incorrect declaration: `NullityBuilder.a` must actually be the moral equivalent of `@o.j.a.Nullable @javax.validation.constraints.NotNull String`, comparable to Checker Framework's `@org.checkerframework.checker.nullness.qual.MonotonicNonNull`. This is because the builder is stateful and cannot start out in a state of `NullityBuilder.a` being non-null. A knock-on consequence is that the builder _must_ validate that the values passed to the record's `@NonNull` components are in fact non-null. Checker Framework makes this extra difficult by rejecting `Objects.requireNonNull`.
   * This is the subtle error.
   * It is reasonable, and perhaps technically necessary, but not truly important, that the `NullityBuilder a(@NonNull String a)` receiver had the `@NonNull` annotation propagated. It is _reasonable_ because passing a `@Nullable` value is clearly nonsensical but it is _unimportant_ because `NullityBuilder.a` cannot practically be `@NonNull`.
   * Arguably `NullityBuilder.a` and all other `@NonNull` record components could be promoted to parameters on the `builder()` method. This would undermine the utility of the resulting builder, however.
   * See also <https://checkerframework.org/manual/#initialization-warnings-constructor>

POM:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>example</groupId>
  <artifactId>avaje-record-builder-nullity</artifactId>
  <version>1.0-SNAPSHOT</version>
  <name>avaje-record-builder-nullity</name>

  <properties>
    <checker.version>3.42.0</checker.version>
    <maven-compiler-plugin.version>3.11.0</maven-compiler-plugin.version>
    <maven.compiler.release>17</maven.compiler.release>
    <record.builder.version>1.0</record.builder.version>
    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    <maven.compiler.source>17</maven.compiler.source>
    <maven.compiler.target>17</maven.compiler.target>
  </properties>

  <dependencies>
    <dependency>
      <groupId>io.avaje</groupId>
      <artifactId>avaje-record-builder</artifactId>
      <version>${record.builder.version}</version>
      <scope>provided</scope>
    </dependency>
    <dependency>
      <groupId>org.jspecify</groupId>
      <artifactId>jspecify</artifactId>
      <version>0.3.0</version>
    </dependency>
    <dependency>
      <groupId>org.checkerframework</groupId>
      <artifactId>checker</artifactId>
      <version>${checker.version}</version>
    </dependency>
    <dependency>
      <groupId>org.checkerframework</groupId>
      <artifactId>checker-qual</artifactId>
      <version>${checker.version}</version>
    </dependency>
    <dependency>
      <groupId>org.checkerframework</groupId>
      <artifactId>checker-util</artifactId>
      <version>${checker.version}</version>
    </dependency>
  </dependencies>

  <build>
    <plugins>
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>${maven-compiler-plugin.version}</version>
        <configuration>
          <compilerArgs>
          </compilerArgs>
          <annotationProcessorPaths>
            <path>
              <groupId>io.avaje</groupId>
              <artifactId>avaje-record-builder</artifactId>
              <version>${record.builder.version}</version>
            </path>
          </annotationProcessorPaths>
          <annotationProcessors>
            <annotationProcessor>io.avaje.recordbuilder.internal.RecordProcessor</annotationProcessor>
          </annotationProcessors>
        </configuration>
      </plugin>
    </plugins>
  </build>

  <profiles>
    <profile>
      <id>checkerframework</id>
      <activation>
        <jdk>[9,)</jdk>
      </activation>
      <build>
        <plugins>
          <plugin>
            <artifactId>maven-compiler-plugin</artifactId>
            <executions>
              <execution>
                <id>default-compile</id>
                <configuration>
                  <fork>true</fork>
                  <annotationProcessorPaths combine.children="append">
                    <path>
                      <groupId>org.checkerframework</groupId>
                      <artifactId>checker</artifactId>
                      <version>${checker.version}</version>
                    </path>
                  </annotationProcessorPaths>
                  <annotationProcessors combine.children="append">
                    <annotationProcessor>org.checkerframework.checker.nullness.NullnessChecker</annotationProcessor>
                  </annotationProcessors>
                  <compilerArgs combine.children="append">
                    <arg>-Xmaxerrs</arg>
                    <arg>10000</arg>
                    <arg>-Xmaxwarns</arg>
                    <arg>10000</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.code=ALL-UNNAMED</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.processing=ALL-UNNAMED</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED</arg>
                    <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
                    <arg>-J--add-opens=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED</arg>
                  </compilerArgs>
                </configuration>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>
    </profile>
  </profiles>

</project>
```

### PR DESCRIPTION
**Pull request by SentryMan**
- all builder fields will generate with a Jspecify @Nullable annotation
- if any form of `@NonNull` is detected on a component, the generated `build` method will validate that the value isn't null
- adds support for Jspecify `@NullMarked` and `@NullUnmarked`

Resolves #43 (At least for the checker framework)

also has changes from #45

### CODE REVIEW

**Comment by SentryMan** [comment]
Alright, we're now good with both checker framework and nullaway.

**Comment by commonquail** [comment]
Yep, Just Works™. Very cool, thank you. :)

File: avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java [hunk 1]
```diff
@@ -20,43 +20,52 @@ static String classTemplate(
...
 		      * Return a new {2} instance with all fields set to the current values in this builder
 		      */
 		     public {2}{10} build() '{'
-		         return new {2}{10}({7});
+		       return new {2}{10}({7});
+		     '}'
+
+		     private static <T> T requireNonNull(@Nullable T obj) '{'
+		       if (obj == null) '{'
+		         throw new IllegalStateException(\"{2}Builder expects all nonnull values to not be null \");
```
Comment (commonquail) [inline avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java:62]: The only remark I have is that this error message is a non-issue in any null-analysis context, _however_, in the default operation (where null can propagate freely) it is difficult to debug:

```
jshell> NullityBuilder.builder().build()
|  Exception java.lang.IllegalStateException: NullityBuilder expects all nonnull values to not be null
|        at NullityBuilder.requireNonNull (#5:38)
|        at NullityBuilder.build (#5:33)
|        at (#7:1)
```

There can be multiple `requireNonNull` statements in the same line so the line number in the stack trace is not much of a help. Sadly I can think of no easy or gratis solve for that -- only passing in the name of the field.

**Review by commonquail — APPROVED** [review]

File: avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java (hunk 1 above)
Comment (SentryMan) [reply avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/Templates.java:62]: > only passing in the name of the field.

hey that's a pretty good idea

File: avaje-record-builder/src/main/java/io/avaje/recordbuilder/RecordBuilder.java [hunk 2]
```diff
@@ -3,38 +3,76 @@
...
+    /** What nullable annotation to use for the builder fields. */
+    Class<? extends Annotation> nullableAnnotation() default Nullable.class;
+
+    /** The interfaces the generated builder will extend. */
+    Class<?>[] builderInterfaces() default {};
+  }
+
+  @Retention(SOURCE)
+  @Target({TYPE, PACKAGE, MODULE})
+  @interface GlobalConfig {
```
Comment (rob-bygrave) [inline avaje-record-builder/src/main/java/io/avaje/recordbuilder/RecordBuilder.java:63]: In terms of supporting partial compilation this state generally needs to be stored and retrieved. For example, a partial compile rebuilds a builder but does not include the package etc that GlobalConfig is on.

That is, I didn't see any code that stores this state and reloads this state - so I don't think this supports a partial compile that doesn't include the GlobalConfig.

**Comment by rob-bygrave** [comment]
So, I'm thinking this maybe could work better / more accurately in that with the current `enforceNullSafety` it pretty much assuming that enforceNullSafety = true the record is in a `@NullMarked` scope [with the scope based on class then package then module etc annotated with `@NullMarked`].

That is, if code is using jspecify then there is `@NullMarked` and `@NullUnmarked` on the module/package/class to control the scope of "it should be NonNull or Null unspecified by default". This is PR currently working / passing with the assumption that `enforceNullSafety = true` matches when the record is in a `@NullMarked` scope.

That is, currently there isn't code in the annotation processor to traverse up the enclosed types (class -> package -> module) looking for a `@NullMarked` or `@NullUnmarked` and using that to determine the NonNull'ness of a component of a record (and instead the code just uses the enforceNullSafety flag and assumes that matches what the jspecify scope has defined).

I'm pondering if it would be better if the enforceNullSafety was removed completely and instead if a record component isn't explicitly defined as NonNull or Nullable then:
- Traverse up to the record type and look for `@NullMarked` or `@NullUnmarked`, if neither are present then ...
- Traverse up to the package and look for `@NullMarked` or `@NullUnmarked`, if neither are present then ...
- Traverse up to the parent package and look for `@NullMarked` or `@NullUnmarked`, if neither are present then ...
- ... (until we get to the top package)
- Traverse to the module and look for `@NullMarked` or `@NullUnmarked`

**Comment by SentryMan** [comment]
> I'm pondering if it would be better

why not both?

**Comment by SentryMan** [comment]
We can probably also try to annotation process those nullmark annotations

in any case we can add that as a fast followup, this PR is getting pretty big.

**Comment by commonquail** [comment]
> So, I'm thinking this maybe could work better / more accurately [...]

Yeah, that sounds right. Checker Framework presently supports neither annotation and NullAway has only experimental support for them (and the default happens to be acceptable for my use cases) so I have not experimented with those combinations.

**Comment by SentryMan** [comment]
> > why not both?
> 
> I suppose because...

I'm thinking more along the lines of somebody who doesn't use static null analysis and wants the builders to enforce non-null at runtime anyway.

**Comment by SentryMan** [comment]
> Traverse up to the record type and look for @NullMarked or @NullUnmarked, if neither are present then .

hmm, yeah wasn't that hard to add

**Comment by rbygrave** [comment]
I'm not keen on aspects of this PR because it adds a bunch of features that I'm not sure we want long term.

- `enforceNullSafety()` I don't think this is justified, I think people should just use JSpecify
- `builderInterfaces` I haven't seen any justification for this yet
- GlobalSettings - I'd prefer if we can live without this
- We might even look to only support JSpecify annotations

How do we feel about those features?

**Comment by SentryMan** [comment]
@rbygrave please review when you get time

**Comment by SentryMan** [comment]
@rbygrave please review and deploy an RC when you get time

File: avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordBuilderPrism.java [hunk 3]
```diff
@@ -160,13 +162,38 @@ static final class Values {
...
      */
     AnnotationValue getters() {
       return values.get("getters");
     }
+
+    /**
+     * Return the AnnotationValue corresponding to the enforceNullSafety() member of the annotation,
+     * or null when the default value is implied.
+     */
+    AnnotationValue enforceNullSafety() {
```
Comment (rbygrave) [inline avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordBuilderPrism.java:178]: Rob to check this after merge.

File: avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordBuilderPrism.java [hunk 4]
```diff
@@ -160,13 +162,38 @@ static final class Values {
...
+     */
+    AnnotationValue nullableAnnotation() {
+      return values.get("nullableAnnotation");
+    }
+
+    /**
+     * Return the AnnotationValue corresponding to the builderInterfaces() member of the annotation,
+     * or null when the default value is implied.
+     */
+    AnnotationValue builderInterfaces() {
```
Comment (rbygrave) [inline avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordBuilderPrism.java:194]: Rob to check this after merge.

File: avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordModel.java [hunk 5]
```diff
@@ -36,6 +36,11 @@ final class RecordModel {
     this.components = components;
     importTypes.add("io.avaje.recordbuilder.Generated");
     importTypes.add("java.util.function.Consumer");
+    importTypes.add("java.util.function.Consumer");
```
Comment (rbygrave) [inline avaje-record-builder-core/src/main/java/io/avaje/recordbuilder/internal/RecordModel.java:39]: Double import line. Rob to check this after merge.

**Pull request merged**